src/Compilers/Core/Desktop/SimpleAnalyzerAssemblyLoader.cs

```diff
@@ -0,0 +1,193 @@
+        ////{
+        ////    if (analyzerPaths == null)
+        ////    {
+        ////        throw new ArgumentNullException(nameof(analyzerPaths));
+        ////    }
+
+        ////    _dependencyPaths = ImmutableArray.CreateRange(analyzerPaths);
+        ////}
+
+        public void AddDependencyLocation(string fullPath)
+        {
+            if (fullPath == null)
+            {
+                throw new ArgumentNullException(nameof(fullPath));
+            }
+
+            if (!_dependencyPaths.Contains(fullPath, StringComparer.OrdinalIgnoreCase))
+            {
+                _dependencyPaths.Add(fullPath);
+            }
+        }
+
+        /// <summary>
+        /// Given a path to an assembly on disk, loads and returns the
+        /// corresponding <see cref="Assembly"/> object.
+        /// </summary>
+        /// <remarks>
+        /// Multiple calls with the same path will return the same 
+        /// <see cref="Assembly"/> instance.
+        /// 
+        /// Also hooks the <see cref="AppDomain.AssemblyResolve"/> event in
+        /// case it needs to load a dependency.
+        /// </remarks>
+        public Assembly LoadFromPath(string fullPath)
+        {
+            CompilerPathUtilities.RequireAbsolutePath(fullPath, nameof(fullPath));
+
+            lock (_guard)
+            {
+                Assembly assembly;
+                if (_pathsToAssemblies.TryGetValue(fullPath, out assembly))
+                {
+                    return assembly;
+                }
+
+                assembly = LoadCore(fullPath);
+
+                if (!_hookedAssemblyResolve)
+                {
+                    _hookedAssemblyResolve = true;
+
+                    AppDomain.CurrentDomain.AssemblyResolve += CurrentDomain_AssemblyResolve;
+                }
+
+                return assembly;
+            }
+        }
+
+        private Assembly LoadCore(string fullPath)
+        {
+            Assembly assembly = Assembly.LoadFrom(fullPath);
+            string assemblyName = assembly.FullName;
+
+            _pathsToAssemblies[fullPath] = assembly;
+            _namesToAssemblies[assemblyName] = assembly;
+
+            return assembly;
+        }
+
+        private Assembly CurrentDomain_AssemblyResolve(object sender, ResolveEventArgs args)
+        {
+            string requestedNameWithPolicyApplied = AppDomain.CurrentDomain.ApplyPolicy(args.Name);
+
+            lock (_guard)
+            {
+                Assembly assembly;
+                if (_namesToAssemblies.TryGetValue(requestedNameWithPolicyApplied, out assembly))
+                {
+                    return assembly;
+                }
+
+                AssemblyIdentity requestedAssemblyIdentity;
+                if (!AssemblyIdentity.TryParseDisplayName(requestedNameWithPolicyApplied, out requestedAssemblyIdentity))
+                {
+                    return null;
+                }
+
+                foreach (string candidatePath in _dependencyPaths)
```

Comment:
Would be nice to be able to exit the lock here. Take a snapshot of "_dependencyPaths"?